js/base-tool-table.js

```diff
@@ -0,0 +1,393 @@
+	
+	// A function for generating and saving tables from JSONs of the Roll20 format
+	function createR20Format(t) {
+		const r20t = d20.Campaign.rollabletables.create({
+			name: t.name.replace(/\s+/g, "-"),
+			showplayers: t.isShown,
+			id: d20plus.ut.generateRowId(),
+		});
+		
+		r20t.tableitems.reset(t.items.map(i => {
+			const out = {
+				id: d20plus.ut.generateRowId(),
+				name: i.row,
+			};
+			if (i.weight !== undefined) out.weight = i.weight;
+			if (i.avatar) out.avatar = i.avatar;
+			return out;
+		}));
+		r20t.tableitems.forEach(it => it.save());
+	}
+	
+	// Function to create from create your own format
+	// Note that due to the simplicity of the format, hiding from players, weight, and avatars can't be configured
+	// Returns true on success, false on failure
+	function createCreateFormat(t) {
+		// Split the input into lines
+		const lines = t.split("\n");
+		
+		// Ensure there's a table name and at least one element
+		if (lines.length < 2) {
+			return [false, "Must contain table name and at least one element"];
+		}
+		
+		// Create table object, using lines[0] as the name
+		const r20t = d20.Campaign.rollabletables.create({
+			name: lines[0].replace(/\s+/g, "-"),
+			showplayers: true,
+			id: d20plus.ut.generateRowId(),
+		});
+		
+		// Create a row from each line
+		r20t.tableitems.reset(lines.slice(1).map(i => {
+			const out = {
+				id: d20plus.ut.generateRowId(),
+				name: i,
+			};
+			return out;
+		}));
+		
+		r20t.tableitems.forEach(it => it.save());
+
+		return [true, null];
+	}
+	
+	// Simplified version of getFromPaste which checks if the paste is valid
+	function validatePaste(paste) {
+		let valid = true;
+		let error = null;
+		let tbl = false;
+		
+		paste.split("\n").forEach(line => parseLine(line.trim()));
+		parseLine(""); // ensure trailing newline
+		return [valid, error];
+		
+		function parseLine(line) {
+			// Check to see if already invalid
+			if (!valid) return;
+			
+			// Check to see if current line imports an item
+			if (line.startsWith("!import-table-item")) {
+				if (!tbl) {
+					valid = false;
+					error = "No !import-table statement found";
+				}
+			}
+			// Check to see if current line imports a table
+			else if (line.startsWith("!import-table")) {
+				if (tbl) {
+					valid = false;
+					error = "No blank line found between tables"
+				}
+				tbl = true;
+			}
+			// Check to ensure against invalid line
+			else if (line.trim()) {
+				valid = false;
+				error = "Non-empty line which didn't match !import-table or !import-table-item";
+			}
+			// Allow multiple tables
+			else {
+				tbl = false;
+			}
+		}
+	}
+	
+	function getFromPaste(paste) {
+		const tables = [];
+		let tbl = null;
+		
+		paste.split("\n").forEach(line => parseLine(line.trim()));
+		parseLine(""); // ensure trailing newline
+		return tables;
+		
+		function parseLine(line) {
+			if (line.startsWith("!import-table-item")) {
+				if (!tbl) {
+					throw new Error("No !import-table statement found");
```

Comment:
this function is out of line with the others--it could `return {tables, error}` instead of throwing errors?

As an aside, this could _also_ then allow partial parses; if we have some valid `tables` at the start of the paste, we could import those, at least. This would require some more UX tinkering, though, to prevent the good tables from being added multiple times if the user then fixed their paste and re-ran. Not something we need to do right now, though

Reply:
Don't want to do partial imports as that could cause confusion.

Reply:
Good with me!